storage/atsd/atsd.go

```diff
@@ -0,0 +1,400 @@
+		atsdHttp.NewMetric(containerFilesystemWritesMerged).SetDataType(atsdHttp.LONG),
+		atsdHttp.NewMetric(containerFilesystemWriteTime).SetDataType(atsdHttp.LONG),
+
+		atsdHttp.NewMetric(containerCpuUsageSystemPct).SetDataType(atsdHttp.FLOAT),
+		atsdHttp.NewMetric(containerCpuUsageTotalPct).SetDataType(atsdHttp.FLOAT),
+		atsdHttp.NewMetric(containerCpuUsageUserPct).SetDataType(atsdHttp.FLOAT),
+
+		atsdHttp.NewMetric(containerCpuHostUsageSystemPct).SetDataType(atsdHttp.FLOAT),
+		atsdHttp.NewMetric(containerCpuHostUsageTotalPct).SetDataType(atsdHttp.FLOAT),
+		atsdHttp.NewMetric(containerCpuHostUsageUserPct).SetDataType(atsdHttp.FLOAT),
+		atsdHttp.NewMetric(containerCpuUsagePerCpuPct).SetDataType(atsdHttp.FLOAT),
+	}
+
+	for _, metric := range metrics {
+		self.innerStorage.RegisterMetric(metric)
+	}
+}
+
+func (self *Storage) AddStats(ref info.ContainerReference, stats *info.ContainerStats) error {
+	timestamp := uint64(stats.Timestamp.UnixNano())
+
+	if isEnabledToStore(self.machineName, ref, self.userCgroupsEnabled, stats) {
+		if self.needToSendSeries(ref.Name, timestamp) {
+
+			cpuSeriesCommands := CpuSeriesCommandsFromStats(self.machineName, ref, stats)
+			ioSeriesCommands := IOSeriesCommandsFromStats(self.machineName, ref, self.includeAllMajorNumbers, stats)
+			memorySeriesCommands := MemorySeriesCommandsFromStats(self.machineName, ref, stats)
+			taskSeriesCommands := TaskSeriesCommandsFromStats(self.machineName, ref, stats)
+			networkSeriesCommands := NetworkSeriesCommandsFromStats(self.machineName, ref, stats)
+			fileSystemSeriesCommands := FileSystemSeriesCommandsFromStats(self.machineName, ref, stats)
+
+			derivedCpuSeries := CalculateDerivedSeriesCpuCommands(self.machineName+ref.Name, stats)
+
+			self.innerStorage.SendSeriesCommands(cpuGroup, cpuSeriesCommands)
+			self.innerStorage.SendSeriesCommands(cpuGroup, derivedCpuSeries)
+			self.innerStorage.SendSeriesCommands(ioGroup, ioSeriesCommands)
+			self.innerStorage.SendSeriesCommands(memoryGroup, memorySeriesCommands)
+			self.innerStorage.SendSeriesCommands(taskGroup, taskSeriesCommands)
+			self.innerStorage.SendSeriesCommands(networkGroup, networkSeriesCommands)
+			self.innerStorage.SendSeriesCommands(filesytemGroup, fileSystemSeriesCommands)
+			self.updateSeriesSentTime(ref.Name, timestamp)
+		}
+
+		if self.needToSendProperties(ref.Name, timestamp) {
+			properties := RefToPropertyCommands(self.machineName, ref, timestamp)
+			self.innerStorage.SendPropertyCommands(properties)
+			self.updatePropertySentTime(ref.Name, timestamp)
+		}
+
+		if self.needToSendEntityTags(ref.Name) {
+			entities := RefToEntityCommands(self.machineName, ref)
+			self.innerStorage.SendEntityTagCommands(entities)
+			self.updateEntitySentStatus(ref.Name)
+		}
+	}
+	return nil
+}
+
+func (self *Storage) Close() error {
+	return nil
+}
+
+func (self *Storage) needToSendEntityTags(containerRefName string) bool {
+	isFirst, ok := self.firstTime[containerRefName]
```

Comment:
Invert the firstTime logic (e.g. `hasSentEntityTags`), then you don't need to check ok, just: `return !hasSentEntityTags[containerRefName]`